File: e2e/chromedp/auth_test.go
package e2e

import (
	"context"
	"testing"
	"time"

	"github.com/chromedp/chromedp"
)

// TestUnauthenticatedRedirectsToLocalLogin は
// e2e/playwright/tests/auth.spec.ts の同名シナリオ(1つ目)に対応する。
func TestUnauthenticatedRedirectsToLocalLogin(t *testing.T) {
	ctx, cancel := context.WithTimeout(newBrowserContext(t), 30*time.Second)
	defer cancel()

	// /tasksは保護ルート。useAuthが叩く/api/meの401をapiFetchが検知し、
	// /login?redirect=... へ遷移する(Keycloakへの自動遷移はしない。CONTRACT.md セクション16)
	err := chromedp.Run(ctx,
		chromedp.Navigate(frontendBaseURL()+"/tasks"),
		chromedp.WaitVisible(`[data-testid="login-email-input"]`, chromedp.ByQuery),
	)
	if err != nil {
		t.Fatalf("未ログイン状態でのリダイレクトに失敗: %v", err)
	}
}

// TestLocalHMACLoginAndLogout は
// 「ローカル認証(HMAC版・既定)でログイン→タスク一覧表示→ログアウト→ログイン画面に戻る」を検証する。
func TestLocalHMACLoginAndLogout(t *testing.T) {
	ctx, cancel := context.WithTimeout(newBrowserContext(t), 30*time.Second)
	defer cancel()

	if err := chromedp.Run(ctx, chromedp.Navigate(frontendBaseURL()+"/tasks")); err != nil {
		t.Fatalf("Navigate() error = %v", err)
	}
	if err := loginViaLocal(ctx, false); err != nil {
		t.Fatalf("loginViaLocal(HMAC) error = %v", err)
	}

	// ローカルセッションのログアウトはKeycloakのRP-Initiated Logoutを経由せず、
	// bffがFrontendBaseURLへの相対パスをそのまま返す(CONTRACT.md セクション16.4)
	// window.location.hrefでの遷移後、未ログイン状態の/がuseAuthの401検知で
	// 再度/loginへリダイレクトされる、という2段の自動遷移になる。
	logoutCtx, cancelLogout := context.WithTimeout(ctx, 15*time.Second)
	defer cancelLogout()
	err := chromedp.Run(logoutCtx,
		chromedp.Click(`[data-testid="logout-button"]`, chromedp.ByQuery),
		chromedp.WaitVisible(`[data-testid="login-email-input"]`, chromedp.ByQuery),
	)
	if err != nil {
		t.Fatalf("ログアウト後にログイン画面へ戻らなかった: %v", err)
	}
}

// TestLocalRSALogin は「ローカル認証(RSA版)でログイン→タスク一覧表示」を検証する。
func TestLocalRSALogin(t *testing.T) {
	ctx, cancel := context.WithTimeout(newBrowserContext(t), 30*time.Second)
	defer cancel()

	if err := chromedp.Run(ctx, chromedp.Navigate(frontendBaseURL()+"/tasks")); err != nil {
		t.Fatalf("Navigate() error = %v", err)
	}
	if err := loginViaLocal(ctx, true); err != nil {
		t.Fatalf("loginViaLocal(RSA) error = %v", err)
	}
}

// TestLocalLoginWrongPassword は誤ったパスワードでのログイン失敗を検証する。
func TestLocalLoginWrongPassword(t *testing.T) {
	ctx, cancel := context.WithTimeout(newBrowserContext(t), 30*time.Second)
	defer cancel()

	err := chromedp.Run(ctx,
		chromedp.Navigate(frontendBaseURL()+"/login"),
		chromedp.WaitVisible(`[data-testid="login-email-input"]`, chromedp.ByQuery),
		chromedp.SendKeys(`[data-testid="login-email-input"]`, "local-user@example.com", chromedp.ByQuery),
		chromedp.SendKeys(`[data-testid="login-password-input"]`, "wrong-password", chromedp.ByQuery),
		chromedp.Click(`[data-testid="login-submit-button"]`, chromedp.ByQuery),
		chromedp.WaitVisible(`[data-testid="login-error"]`, chromedp.ByQuery),
		chromedp.WaitVisible(`[data-testid="login-email-input"]`, chromedp.ByQuery),
	)
	if err != nil {
		t.Fatalf("誤ったパスワードでのエラー表示に失敗: %v", err)
	}
}

// TestKeycloakLoginAndLogout は
// 「Keycloakでログイン→タスク一覧表示→ログアウト→ログイン画面に戻る」を検証する。
func TestKeycloakLoginAndLogout(t *testing.T) {
	ctx, cancel := context.WithTimeout(newBrowserContext(t), 30*time.Second)
	defer cancel()

	if err := chromedp.Run(ctx, chromedp.Navigate(frontendBaseURL()+"/tasks")); err != nil {
		t.Fatalf("Navigate() error = %v", err)
	}
	if err := loginViaKeycloak(ctx); err != nil {
		t.Fatalf("loginViaKeycloak() error = %v", err)
	}

	// ログアウト後の遷移は複数段ある: bffの/api/auth/logout(fetch)
	// → window.location.hrefでKeycloakのend_session_endpointへ実遷移(SSOセッション終了)
	// → post_logout_redirect_uriで frontendのトップへ戻る
	// → ProtectedLayoutが未ログインを検知し /login へ自動的にリダイレクトされる
	// 【playwright版と同じ理由で、中間URLは待たず最終状態だけを十分な猶予時間で待つ】
	// 途中のURLをwaitForで待つと、その後さらに自動リダイレクトが続いている最中に
	// 次のアクションが割り込んでしまい、ログイン画面に正しくたどり着けないことがある。
	logoutCtx, cancelLogout := context.WithTimeout(ctx, 15*time.Second)
	defer cancelLogout()
	err := chromedp.Run(logoutCtx,
		chromedp.Click(`[data-testid="logout-button"]`, chromedp.ByQuery),
		chromedp.WaitVisible(`[data-testid="login-email-input"]`, chromedp.ByQuery),
	)
	if err != nil {
		t.Fatalf("Keycloakログアウト後にログイン画面へ戻らなかった: %v", err)
	}
}

File: e2e/selenium/test/auth.test.js
const assert = require("assert");
const { By, until } = require("selenium-webdriver");
const { buildDriver, testId, loginViaKeycloak, loginViaLocal, BASE_URL } = require("./helpers");

describe("認証(BFFパターン)", function () {
  let driver;

  beforeEach(async function () {
    driver = await buildDriver();
  });

  afterEach(async function () {
    await driver.quit();
  });

  it("未ログイン状態で/tasksへアクセスするとローカルログイン画面(HMAC版・既定)へリダイレクトされる", async function () {
    // /tasksは保護ルート
    // useAuthが叩く/api/meの401をapiFetchが検知し、/login?redirect=... へ遷移する
    // (Keycloakへの自動遷移はしない CONTRACT.md セクション16)
    await driver.get(`${BASE_URL}/tasks`);
    await driver.wait(until.elementLocated(testId("login-email-input")), 10000);
    const el = await driver.findElement(testId("login-email-input"));
    assert.ok(await el.isDisplayed());
  });

  it("ローカル認証(HMAC版・既定)でログイン→タスク一覧表示→ログアウト→ログイン画面に戻る", async function () {
    await driver.get(`${BASE_URL}/tasks`);
    await loginViaLocal(driver);
    const list = await driver.findElement(testId("task-list"));
    assert.ok(await list.isDisplayed());

    await driver.findElement(testId("logout-button")).click();
    await driver.wait(until.elementLocated(testId("login-email-input")), 15000);
  });

  it("ローカル認証(RSA版)でログイン→タスク一覧表示", async function () {
    await driver.get(`${BASE_URL}/tasks`);
    await loginViaLocal(driver, true);
    const list = await driver.findElement(testId("task-list"));
    assert.ok(await list.isDisplayed());
  });

  it("誤ったパスワードでのログインはエラーメッセージを表示し、ログイン画面に留まる", async function () {
    await driver.get(`${BASE_URL}/login`);
    await driver.wait(until.elementLocated(testId("login-email-input")), 10000);
    await driver.findElement(testId("login-email-input")).sendKeys("local-user@example.com");
    await driver.findElement(testId("login-password-input")).sendKeys("wrong-password");
    await driver.findElement(testId("login-submit-button")).click();

    await driver.wait(until.elementLocated(testId("login-error")), 10000);
    const emailInput = await driver.findElement(testId("login-email-input"));
    assert.ok(await emailInput.isDisplayed());
  });

  it("Keycloakでログイン→タスク一覧表示→ログアウト→ログイン画面に戻る", async function () {
    await driver.get(`${BASE_URL}/tasks`);
    await loginViaKeycloak(driver);
    const list = await driver.findElement(testId("task-list"));
    assert.ok(await list.isDisplayed());

    // ログアウト後は、
    // bff の /api/auth/logout(非同期fetch) →
    // Keycloak の end_session_endpointへの実遷移 →
    // post_logout_redirect_uriで5173へ戻る →
    // ProtectedLayoutが未ログインを検知し /login へ自動的に遷移する
    // という複数段の自動リダイレクトが発生する(Playwright実行時に判明した挙動と同じ)
    //
    // ここで明示的にdriver.get()すると、この自動リダイレクトの途中に割り込んでしまい不安定になるため、
    // クリック後は一連のリダイレクトが収束した最終状態
    // (ローカルログイン画面)を十分な猶予時間で待つだけにする
    await driver.findElement(testId("logout-button")).click();
    await driver.wait(until.elementLocated(testId("login-email-input")), 15000);
  });
});
